File: src/lib/productivityCalculation.ts
// Productivity calculation based on Python script logic
// Computes Area, ETa, TBP and Productivity tables for each season

export interface CropRecord {
  year: number;
  season: number; // 0=Annual, 1=Kharif, 2=Rabi, 3=Zaid, 4=Full year
  cropType: string;
  area: number;
  eta: number;
  tbp: number;
  status: string;
  productivity?: number;
}

export interface SeasonTables {
  area: Record<string, Record<string, number>>; // year -> cropType -> value
  eta: Record<string, Record<string, number>>;
  tbp: Record<string, Record<string, number>>;
  productivity: Record<string, Record<string, number>>;
  weightedProductivity: Record<string, number>; // year -> weighted avg
  years: number[];
  crops: string[];
  averageWeightedProductivity: number;
}

export interface ProductivityData {
  summary: Array<{
    year: number | string;
    kharif: number | null;
    rabi: number | null;
    zaid: number | null;
    annual: number | null;
  }>;
  average: {
    kharif: number;
    rabi: number;
    zaid: number;
    annual: number;
  };
  seasonTables: {
    kharif: SeasonTables | null;
    rabi: SeasonTables | null;
    zaid: SeasonTables | null;
    annual: SeasonTables | null;
  };
}

export const PRODUCTIVITY_SEASON_MAP: Record<number, string> = {
  0: 'annual',
  1: 'kharif',
  2: 'rabi',
  3: 'zaid',
  4: 'fullYear'
};

export const PRODUCTIVITY_SEASON_OPTIONS = [
  { value: 'kharif', label: 'Kharif', code: 1 },
  { value: 'rabi', label: 'Rabi', code: 2 },
  { value: 'zaid', label: 'Zaid', code: 3 },
  { value: 'annual', label: 'Annual', code: 0 },
];

function parseProductivityCSV(csvText: string): CropRecord[] {
  const lines = csvText.trim().split('\n');
  if (lines.length < 2) return [];

  const header = lines[0].split(',').map(h => h.trim());
  const getIndex = (name: string) => header.findIndex(h => h.toLowerCase() === name.toLowerCase());

  const yearIdx = getIndex('year');
  const seasonIdx = getIndex('Season');
  const cropTypeIdx = getIndex('Crop Type');
  const areaIdx = getIndex('Area');
  const etaIdx = getIndex('ETa');
  const tbpIdx = getIndex('TBP');
  const statusIdx = getIndex('status');

  const records: CropRecord[] = [];

  for (let i = 1; i < lines.length; i++) {
    const values = lines[i].split(',');
    if (values.length < header.length) continue;

    const year = parseInt(values[yearIdx]);
    const season = parseInt(values[seasonIdx]);
    const area = parseFloat(values[areaIdx]);
    const eta = parseFloat(values[etaIdx]);
    const tbp = parseFloat(values[tbpIdx]);

    if (isNaN(year) || isNaN(season) || isNaN(area) || isNaN(eta) || isNaN(tbp)) continue;

    const productivity = eta > 0 ? tbp / (eta * 10) : 0;

    records.push({
      year,
      season,
      cropType: values[cropTypeIdx]?.trim() || 'Unknown',
      area,
      eta,
      tbp,
      status: values[statusIdx]?.trim() || '',
      productivity,
    });
  }

  return records;
}

function generateSeasonTables(records: CropRecord[], seasonCode: number): SeasonTables | null {
  // Filter for season and irrigated status
  const base = records.filter(r => r.season === seasonCode && r.status === 'IRRIGATED');
  if (base.length === 0) return null;

  const filteredAvg = base.filter(r => r.area > 0);
  
  const years = [...new Set(base.map(r => r.year))].sort((a, b) => a - b);
  const crops = [...new Set(base.map(r => r.cropType))].sort();

  // 1) AREA (SUM)
  const area: Record<string, Record<string, number>> = {};
  for (const year of years) {
    area[year] = {};
    for (const crop of crops) {
      const sum = base.filter(r => r.year === year && r.cropType === crop)
        .reduce((acc, r) => acc + r.area, 0);
      area[year][crop] = Math.round(sum);
    }
    // Row average
    const rowValues = Object.values(area[year]);
    area[year]['Average'] = Math.round(rowValues.reduce((a, b) => a + b, 0) / rowValues.length);
  }
  // Column averages
  area['Average'] = {};
  for (const crop of [...crops, 'Average']) {
    const colValues = years.map(y => area[y][crop]).filter(v => !isNaN(v));
    area['Average'][crop] = Math.round(colValues.reduce((a, b) => a + b, 0) / colValues.length);
  }

  // 2) ETa (AVERAGE)
  const eta: Record<string, Record<string, number>> = {};
  for (const year of years) {
    eta[year] = {};
    for (const crop of crops) {
      const cropRecords = filteredAvg.filter(r => r.year === year && r.cropType === crop);
      if (cropRecords.length > 0 && area[year][crop] > 0) {
        eta[year][crop] = Math.round(cropRecords.reduce((acc, r) => acc + r.eta, 0) / cropRecords.length);
      } else {
        eta[year][crop] = NaN;
      }
    }
    const rowValues = Object.values(eta[year]).filter(v => !isNaN(v));
    eta[year]['Average'] = rowValues.length > 0 ? Math.round(rowValues.reduce((a, b) => a + b, 0) / rowValues.length) : NaN;
  }
  eta['Average'] = {};
  for (const crop of [...crops, 'Average']) {
    const colValues = years.map(y => eta[y][crop]).filter(v => !isNaN(v));
    eta['Average'][crop] = colValues.length > 0 ? Math.round(colValues.reduce((a, b) => a + b, 0) / colValues.length) : NaN;
  }

  // 3) TBP (AVERAGE)
  const tbp: Record<string, Record<string, number>> = {};
  for (const year of years) {
    tbp[year] = {};
    for (const crop of crops) {
      const cropRecords = filteredAvg.filter(r => r.year === year && r.cropType === crop);
      if (cropRecords.length > 0 && area[year][crop] > 0) {
        tbp[year][crop] = Math.round(cropRecords.reduce((acc, r) => acc + r.tbp, 0) / cropRecords.length);
      } else {
        tbp[year][crop] = NaN;
      }
    }
    const rowValues = Object.values(tbp[year]).filter(v => !isNaN(v));
    tbp[year]['Average'] = rowValues.length > 0 ? Math.round(rowValues.reduce((a, b) => a + b, 0) / rowValues.length) : NaN;
  }
  tbp['Average'] = {};
  for (const crop of [...crops, 'Average']) {
    const colValues = years.map(y => tbp[y][crop]).filter(v => !isNaN(v));
    tbp['Average'][crop] = colValues.length > 0 ? Math.round(colValues.reduce((a, b) => a + b, 0) / colValues.length) : NaN;
  }

  // 4) PRODUCTIVITY (mean + weighted average)
  const productivity: Record<string, Record<string, number>> = {};
  const weightedProductivity: Record<string, number> = {};
  
  for (const year of years) {
    productivity[year] = {};
    for (const crop of crops) {
      const cropRecords = filteredAvg.filter(r => r.year === year && r.cropType === crop);
      if (cropRecords.length > 0) {
        const avg = cropRecords.reduce((acc, r) => acc + (r.productivity || 0), 0) / cropRecords.length;
        productivity[year][crop] = parseFloat(avg.toFixed(2));
      } else {
        productivity[year][crop] = NaN;
      }
    }
    
    // Weighted average for year
    const yearRecords = filteredAvg.filter(r => r.year === year);
    const totalArea = yearRecords.reduce((acc, r) => acc + r.area, 0);
    if (totalArea > 0) {
      const weightedSum = yearRecords.reduce((acc, r) => acc + r.area * (r.productivity || 0), 0);
      weightedProductivity[year] = parseFloat((weightedSum / totalArea).toFixed(2));
      productivity[year]['Average'] = weightedProductivity[year];
    } else {
      weightedProductivity[year] = NaN;
      productivity[year]['Average'] = NaN;
    }
  }

  // Average row
  productivity['Average'] = {};
  for (const crop of crops) {
    const colValues = years.map(y => productivity[y][crop]).filter(v => !isNaN(v));
    productivity['Average'][crop] = colValues.length > 0 
      ? parseFloat((colValues.reduce((a, b) => a + b, 0) / colValues.length).toFixed(2))
      : NaN;
  }
  
  const weightedValues = years.map(y => weightedProductivity[y]).filter(v => !isNaN(v));
  const averageWeightedProductivity = weightedValues.length > 0
    ? parseFloat((weightedValues.reduce((a, b) => a + b, 0) / weightedValues.length).toFixed(2))
    : NaN;
  
  productivity['Average']['Average'] = averageWeightedProductivity;

  return {
    area,
    eta,
    tbp,
    productivity,
    weightedProductivity,
    years,
    crops,
    averageWeightedProductivity,
  };
}

export async function computeProductivityFromCSV(perCropFile?: File): Promise<ProductivityData> {
  let csvText: string;

  if (perCropFile) {
    csvText = await perCropFile.text();
  } else {
    // Default to sample file
    const response = await fetch('/data/stats_IndiaSchemes_perCrop_Ong.csv');
    if (!response.ok) throw new Error('Failed to load productivity data file');
    csvText = await response.text();
  }

  const records = parseProductivityCSV(csvText);

  // Generate tables for each season
  const kharif = generateSeasonTables(records, 1);
  const rabi = generateSeasonTables(records, 2);
  const zaid = generateSeasonTables(records, 3);
  const annual = generateSeasonTables(records, 4); // Season 4 = full year/annual

  // Build summary
  const allYears = new Set<number>();
  [kharif, rabi, zaid, annual].forEach(s => s?.years.forEach(y => allYears.add(y)));
  const sortedYears = [...allYears].sort((a, b) => a - b);

  const summary = sortedYears.map(year => ({
    year,
    kharif: kharif?.weightedProductivity[year] ?? null,
    rabi: rabi?.weightedProductivity[year] ?? null,
    zaid: zaid?.weightedProductivity[year] ?? null,
    annual: annual?.weightedProductivity[year] ?? null,
  }));

  const average = {
    kharif: kharif?.averageWeightedProductivity ?? 0,
    rabi: rabi?.averageWeightedProductivity ?? 0,
    zaid: zaid?.averageWeightedProductivity ?? 0,
    annual: annual?.averageWeightedProductivity ?? 0,
  };

  return {
    summary,
    average,
    seasonTables: { kharif, rabi, zaid, annual },
  };
}

File: python_backend/productivity_calculation.py
"""
Productivity calculation based on Python script logic
Computes Area, ETa, TBP and Productivity tables for each season
"""

import pandas as pd
import numpy as np
from typing import Dict, List, Optional

PRODUCTIVITY_SEASON_MAP = {
    0: 'annual',
    1: 'kharif',
    2: 'rabi',
    3: 'zaid',
    4: 'fullYear'
}


def parse_productivity_csv(csv_path: str) -> pd.DataFrame:
    """Parse productivity CSV file"""
    df = pd.read_csv(csv_path)
    
    # Ensure required columns exist
    required_cols = ['year', 'Season', 'Crop Type', 'Area', 'ETa', 'TBP', 'status']
    for col in required_cols:
        if col not in df.columns:
            raise ValueError(f"Missing required column: {col}")
    
    # Calculate productivity: TBP / (ETa * 10)
    df['productivity'] = df.apply(
        lambda row: row['TBP'] / (row['ETa'] * 10) if row['ETa'] > 0 else 0,
        axis=1
    )
    
    return df


def generate_season_tables(df: pd.DataFrame, season_code: int) -> Optional[Dict]:
    """Generate tables for a specific season"""
    # Filter for season and irrigated status
    base = df[(df['Season'] == season_code) & (df['status'] == 'IRRIGATED')].copy()
    if base.empty:
        return None
    
    filtered_avg = base[base['Area'] > 0].copy()
    
    years = sorted(base['year'].unique())
    crops = sorted(base['Crop Type'].unique())
    
    # 1) AREA (SUM)
    area = {}
    for year in years:
        area[year] = {}
        for crop in crops:
            total = base[(base['year'] == year) & (base['Crop Type'] == crop)]['Area'].sum()
            area[year][crop] = round(total)
        # Row average
        row_values = [area[year][crop] for crop in crops]
        area[year]['Average'] = round(sum(row_values) / len(row_values))
    
    # Column averages
    area['Average'] = {}
    for crop in crops + ['Average']:
        col_values = [area[y][crop] for y in years if crop in area[y]]
        area['Average'][crop] = round(sum(col_values) / len(col_values)) if col_values else 0
    
    # 2) ETa (AVERAGE)
    eta = {}
    for year in years:
        eta[year] = {}
        for crop in crops:
            crop_records = filtered_avg[(filtered_avg['year'] == year) & (filtered_avg['Crop Type'] == crop)]
            if len(crop_records) > 0 and area[year][crop] > 0:
                eta[year][crop] = round(crop_records['ETa'].mean())
            else:
                eta[year][crop] = None
        row_values = [v for v in eta[year].values() if v is not None]
        eta[year]['Average'] = round(sum(row_values) / len(row_values)) if row_values else None
    
    eta['Average'] = {}
    for crop in crops + ['Average']:
        col_values = [eta[y][crop] for y in years if crop in eta[y] and eta[y][crop] is not None]
        eta['Average'][crop] = round(sum(col_values) / len(col_values)) if col_values else None
    
    # 3) TBP (AVERAGE)
    tbp = {}
    for year in years:
        tbp[year] = {}
        for crop in crops:
            crop_records = filtered_avg[(filtered_avg['year'] == year) & (filtered_avg['Crop Type'] == crop)]
            if len(crop_records) > 0 and area[year][crop] > 0:
                tbp[year][crop] = round(crop_records['TBP'].mean())
            else:
                tbp[year][crop] = None
        row_values = [v for v in tbp[year].values() if v is not None]
        tbp[year]['Average'] = round(sum(row_values) / len(row_values)) if row_values else None
    
    tbp['Average'] = {}
    for crop in crops + ['Average']:
        col_values = [tbp[y][crop] for y in years if crop in tbp[y] and tbp[y][crop] is not None]
        tbp['Average'][crop] = round(sum(col_values) / len(col_values)) if col_values else None
    
    # 4) PRODUCTIVITY (mean + weighted average)
    productivity = {}
    weighted_productivity = {}
    
    for year in years:
        productivity[year] = {}
        for crop in crops:
            crop_records = filtered_avg[(filtered_avg['year'] == year) & (filtered_avg['Crop Type'] == crop)]
            if len(crop_records) > 0:
                avg = crop_records['productivity'].mean()
                productivity[year][crop] = round(avg, 2)
            else:
                productivity[year][crop] = None
        
        # Weighted average for year
        year_records = filtered_avg[filtered_avg['year'] == year]
        total_area = year_records['Area'].sum()
        if total_area > 0:
            weighted_sum = (year_records['Area'] * year_records['productivity']).sum()
            weighted_productivity[year] = round(weighted_sum / total_area, 2)
            productivity[year]['Average'] = weighted_productivity[year]
        else:
            weighted_productivity[year] = None
            productivity[year]['Average'] = None
    
    # Average row
    productivity['Average'] = {}
    for crop in crops:
        col_values = [productivity[y][crop] for y in years if crop in productivity[y] and productivity[y][crop] is not None]
        productivity['Average'][crop] = round(sum(col_values) / len(col_values), 2) if col_values else None
    
    weighted_values = [v for v in weighted_productivity.values() if v is not None]
    average_weighted_productivity = round(sum(weighted_values) / len(weighted_values), 2) if weighted_values else None
    productivity['Average']['Average'] = average_weighted_productivity
    
    # Convert year keys to strings for JSON serialization
    area_str = {str(k): v for k, v in area.items()}
    eta_str = {str(k): v for k, v in eta.items()}
    tbp_str = {str(k): v for k, v in tbp.items()}
    productivity_str = {str(k): v for k, v in productivity.items()}
    weighted_productivity_str = {str(k): v for k, v in weighted_productivity.items()}
    
    return {
        'area': area_str,
        'eta': eta_str,
        'tbp': tbp_str,
        'productivity': productivity_str,
        'weighted_productivity': weighted_productivity_str,
        'years': years,
        'crops': crops,
        'average_weighted_productivity': average_weighted_productivity,
    }


def compute_productivity_from_csv(csv_path: str) -> Dict:
    """Compute productivity from CSV file"""
    df = parse_productivity_csv(csv_path)
    
    # Generate tables for each season
    kharif = generate_season_tables(df, 1)
    rabi = generate_season_tables(df, 2)
    zaid = generate_season_tables(df, 3)
    annual = generate_season_tables(df, 4)  # Season 4 = full year/annual
    
    # Build summary
    all_years = set()
    for season_data in [kharif, rabi, zaid, annual]:
        if season_data:
            all_years.update(season_data['years'])
    
    sorted_years = sorted(all_years)
    
    summary = []
    for year in sorted_years:
        summary.append({
            'year': int(year),
            'kharif': kharif['weighted_productivity'].get(str(year)) if kharif else None,
            'rabi': rabi['weighted_productivity'].get(str(year)) if rabi else None,
            'zaid': zaid['weighted_productivity'].get(str(year)) if zaid else None,
            'annual': annual['weighted_productivity'].get(str(year)) if annual else None,
        })
    
    average = {
        'kharif': kharif['average_weighted_productivity'] if kharif else 0,
        'rabi': rabi['average_weighted_productivity'] if rabi else 0,
        'zaid': zaid['average_weighted_productivity'] if zaid else 0,
        'annual': annual['average_weighted_productivity'] if annual else 0,
    }
    
    return {
        'summary': summary,
        'average': average,
        'season_tables': {
            'kharif': kharif,
            'rabi': rabi,
            'zaid': zaid,
            'annual': annual,
        }
    }


if __name__ == '__main__':
    # Test with sample data
    import sys
    if len(sys.argv) > 1:
        result = compute_productivity_from_csv(sys.argv[1])
        print("Summary:", result['summary'])
        print("Average:", result['average'])
